### ISSUE TITLE
Proposal: org-roam successor

### ISSUE THREAD

**Issue by anuvyklack** [meedstrom/org-node#74]
I deeply appreciate your work! Your idea that daily notes are just a sequence of nodes, and the possibility of constructing such sequences based on various criteria, is brilliant!

However, the current implementation is fragmented across the `org-node`, `org-roam`, and `org-node-fakeroam` packages. This fragmentation demands significant effort to bring everything together, which might discourage many potential users. I fear that, as a result, this remarkable package might go unnoticed. Meanwhile, the development of `org-roam` appears stagnant, as even hot bug fixes ([org-roam#2486](https://github.com/org-roam/org-roam/pull/2486)) seem to take an unreasonably long time to merge.

If you don't mind, I would propose you to adopt the necessary parts of the `org-roam` source code, such as the org-roam buffer or any you need (I think the license allows it), and deliver a fully packed, plug-and-play package. I believe it can become a great successor to `org-roam`.

---

Additionally, if you add [org-roam-nursery](https://github.com/chrisbarrett/nursery)-like org dynamic blocks for auto building indexes and [org-super-links](https://github.com/toshism/org-super-links)-like drawers for backlinks — it will be an absolute masterpiece!

**Comment by meedstrom** [comment]
Thank you, I appreciate your excitement!

I agree all that would be awesome. But I find myself in a place where I've done everything I've set out to do, and prefer to just fix bugs.

I could merge the org-roam buffer, that's true, and that'll eliminate `org-node-fakeroam-{fast-render-mode,redisplay-mode,jit-backlinks-mode}`.  Might look into that in January.

Once that's done, you may have less reason to get org-roam and fakeroam.  It would just be if you want to use org-roam's capture system or any of the ecosystem that depends on `org-roam-db-query`. Such as the stuff in that nursery (thanks for the link, some cool stuff there).

And org-super-links drawers -- I want to try them eventually too.

**Comment by real-or-random** [comment]
> I could merge the org-roam buffer, that's true, and that'll eliminate `org-node-fakeroam-{fast-render-mode,redisplay-mode,jit-backlinks-mode}`. Might look into that in January.
> 
> Once that's done, you may have less reason to get org-roam and fakeroam.

Indeed, I think this will be a good middle ground.

**Comment by meedstrom** [comment]
~75% done with a new backlinks buffer, ~400 LoC so far. 

Ideas welcome! Some of mine:

- Call it a "context buffer" because it seems like you could want to put more info in there than just backlinks
- Keep a history so you can go forward-backward with `l` and `r`, like a Help buffer
- Restore point, in contexts that have been rendered before
  - (May be difficult until I understand more of the magit-section API)

**Comment by meedstrom** [comment]
v2.0.0 almost out! If you're antsy, you can preview at:

https://github.com/meedstrom/org-node/tree/volatile

which requires

https://github.com/meedstrom/el-job/tree/dev

But I think I'll have them out within 24 hours anyway.

**Comment by meedstrom** [comment]
It's done.

**Comment by fleimgruber** [comment]
Great job! I just removed all of org-roam from my config. The only thing missing for me is a replacement for https://www.orgroam.com/manual.html#The-roam_002dref-protocol-1. Do you have plans for a similar functionality?

**Comment by meedstrom** [comment]
That's new to me, but why not?

It fits with what I'm coding in recent days, which is turning out to be a way to enable most third-party packages that used to depend on org-roam.   

It's a diskless, throwaway SQLite DB (using the Emacs 29 sqlite.c), and a file fulla shim definitions.

If it works, it'll deprecate org-node-fakeroam altogether.

**Comment by meedstrom** [comment]
Update: went with a slightly different solution: taking a hacksaw to the codebase and spinning out a library. https://github.com/meedstrom/indexed

Once it's  on MELPA, I can shrink org-node by 1/3, focus on UI stuff.  Org-roam would be able to use that library too if they want.

Comments plz. Esp. the RFCs. What's stupid? What's missing?

**Comment by meedstrom** [comment]
v3 on https://github.com/meedstrom/org-node/tree/dev

**Comment by real-or-random** [comment]
> v3 on [`dev`](https://github.com/meedstrom/org-node/tree/dev?rgh-link-date=2025-03-22T13%3A04%3A59.000Z)

I hope I'll find the time to test this to provide some feedback, but can you clarify whether the README is accurate? 

Specifically, I currently use `org-roam`. I assume I should use this:

```
(use-package org-node
  :after org
  :config (org-node-cache-mode))
```

Do you still recommend installing `org-node-fakeroam`? Do you recommend removing `org-roam`?

**Comment by meedstrom** [comment]
Good question, thanks for asking. Now Indexed can write to the org-roam DB, so the main big piece of code left in fakeroam is the patch to speed-up the org-roam buffer, `org-node-fakeroam-fast-render-mode`, but if you use org-node-context buffers instead then you don't need that.  Maybe it's time I (or someone) upstreams the patch...

Guess the package can be removed after that. Just merged a couple of small utilities back into core.

You could try installing like this. Will update the readme.

```elisp
(use-package org-node
  :init
  ;; For org-roam users who find saving slow with
  ;; `org-roam-db-autosync-mode',
  ;; do org-roam's job for it and write to its DB
  (setq org-roam-db-update-on-save nil)
  (setq indexed-roam-overwrite t)
  ;; Use org-roam-capture
  (setq org-node-creation-fn #'org-node-new-via-roam-capture)
  ;; Filenames like org-roam
  (setq org-node-slug-fn #'org-node-slugify-like-roam-default)
  (setq org-node-datestamp-format "%Y%m%d%H%M%S-")
  :config
  (indexed-updater-mode)
  (indexed-roam-mode)
  (org-node-cache-mode))

;; Speed up the org-roam buffer 
;; (NOTE: subject to removal when upstream finally patches this)
;; Tip: You could use org-node's context buffer instead.
(use-package org-node-fakeroam
  :config
  (org-node-fakeroam-fast-render-mode))
```

**Comment by real-or-random** [comment]
I tried the suggested config. I added only `(setq indexed-org-dirs '("~/org/roam"))` (perhaps it should respect `org-roam-directory` if set?). When I try org-node functions, I get an error. I guess a minimal reproducer is `indexed-reset`, which gives me this after a few seconds:
```
Error running timer ‘el-job--poll’: (wrong-number-of-arguments #<subr org-node-fakeroam--remember-roam-dirs> 1)
```

These are the packages (this is doom specific but it should be clear what this does):
```
(package! org-node
  :recipe (:host github :repo "meedstrom/org-node" :branch "dev"))
(package! org-node-fakeroam
  :recipe (:host github :repo "meedstrom/org-node-fakeroam"))
(package! indexed
  :recipe (:host github :repo "meedstrom/indexed"))
```

**Comment by meedstrom** [comment]
Ah. Try without `org-node-fakeroam`. I'll fix it later today. Thanks!

**Comment by real-or-random** [comment]
> Ah. Try without `org-node-fakeroam`. I'll fix it later today. Thanks!

Thanks for the quick response!

I haven't taken a close look, but here are some further issues I encountered so far:
 -  As I said, I have `(setq indexed-org-dirs '("~/org/roam"))` and now some of my files are listed twice in `org-node-list-files`, as `~/org/roam/foo.org` and as `/home/tim/org/roam`). 
 - There are files `.org` listed which are not in `~/org/roam`. Is it possible that any open .org file is also indexed, no matter if it's below a dir in `indexed-org-dirs`? (This would also explain the previous bullet.) Do I need to add other dirs to `indexed-org-dirs-exclude`? If yes, that's not clear from the docs.

**Comment by real-or-random** [comment]
> I have a theory on the doubly-listed files. If you set above option to nil and do `M-x indexed-reset`, does that eliminate them?

No. Neither does setting `(setq indexed-org-dirs '("/home/tim/org/roam"))` instead.

**Comment by meedstrom** [comment]
Could you try `(clrhash indexed--files-to-index)` and reset again? Thanks for testing :)

**Comment by real-or-random** [comment]
> Could you try `(clrhash indexed--files-to-index)` and reset again? Thanks for testing :)

Yep, that did the trick. The files get listed only as `~/...` now, indepedently of whether the dir is configured as `/home/tim/org/roam` or ~/org/roam`.

**Comment by meedstrom** [comment]
Great! As it should be.

If you're testing now, be sure to upgrade el-job + indexed + org-node. Bunch of commits a few hours ago.

I think what's left is mainly doc bugs.

**Comment by real-or-random** [comment]
I had problems with two of my org files which had broken drawers, and I only got this:
```
Error running timer ‘el-job--poll’: (error "File not indexed: ~/org/roam/foo.org")
 ```

Only after fixing the drawers, `indexed-reset` gave me a proper completion message. I guess requiring valid org files is fine, but error handling can be improved. A broken file could just be skipped with a warning.

---

These were the problems:

Missing "END:":
```
:PROPERTIES:
:ID:       dc9da0f9-418c-4922-b5eb-337e500eaff8
:
```

Missing the entire ":END:" line
```
:PROPERTIES:
:CREATED:  [2024-09-20 Fri 14:59]
:END:
```

**Comment by meedstrom** [comment]
Thank you! That reveals an oversight now fixed https://github.com/meedstrom/indexed/commit/22f07d95d02ecf8b5e38a3d2fb8598e23d49f61e. I agree about just skipping.

Now if I butcher a drawer like yours, it handles alright.

**Comment by meedstrom** [comment]
FYI no release the next two weeks, I’ll be computer-free.

Gotta say it bothers me more and more the name ”indexed”, it probably will have to go. And while I’m at it, [the API](https://github.com/meedstrom/indexed/blob/1a0a8e2776cdc861ffbe9b49f99d5aac70c7d89a/indexed.el#L200) is too magical for the sake of shortness. Should overhaul, so accessors will have longer but clearer names.

**Comment by meedstrom** [comment]
No more fragmentation; org-node-fakeroam not needed anymore. That concludes that complaint! ;)

To use org-roam-db based stuff, see [org-mem-roamy-db-mode](https://github.com/meedstrom/org-mem?tab=readme-ov-file#appendix-i-a-sqlite-database-for-free), which is already installed if org-node v3 is.  Packages that use org-roam still pull in org-roam, of course, but you need not enable it.
